### PR TITLE
Just collaborating - Please be nice

### PR DESCRIPTION
**Pull request by christgf**
* Tidy go.mod
* cmd/api: HTTP server timeouts configuration and other minor tweaks
* cmd/api: HTTP server graceful shutdown
* cmd/api: HTTP handle functions now resistant to panics
* cmd/api: HTTP handle functions can (should) now return unhandled errors
* cmd/api: Protecting against accidentally (or maliciously) large requests
* cmd/api: Globalising MIME sniffing protection
* server: HTTP handle functions now return unhandled errors

### CODE REVIEW

File: cmd/api/main.go [hunk 1]
```diff
@@ -1,37 +1,92 @@
...
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() { // recover in case of panics
+			if r := recover(); r != nil {
+				http.Error(w, "service unavailable", http.StatusServiceUnavailable)
+			}
+		}()
+		r.Body = http.MaxBytesReader(w, r.Body, 100*1024*1024) // limit the size of incoming request body
+		w.Header().Set("X-Content-Type-Options", "nosniff")    // no sniffing please
+		if err := handleFunc(w, r); err != nil {
+			http.Error(w, "internal server error", http.StatusInternalServerError)
```
Comment (christgf) [inline cmd/api/main.go]: Unhandled/unrecoverable HTTP processing errors handled here. HTTP500 for now (since this is what individual handlers were doing) but could be something else.

File: cmd/api/main.go [hunk 2]
```diff
@@ -1,37 +1,92 @@
...
+func registerFileServe(router *mux.Router, path string) {
+	router.Handle("/static/", http.StripPrefix("/static", http.FileServer(http.Dir(path))))
+}
+
+func serveHandleFunc(handleFunc func(http.ResponseWriter, *http.Request) error) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() { // recover in case of panics
+			if r := recover(); r != nil {
+				http.Error(w, "service unavailable", http.StatusServiceUnavailable)
+			}
```
Comment (christgf) [inline cmd/api/main.go]: Definitely need some logging here, but should be handled in a different PR. I'm not sure we have enough in terms of logging just yet and trying to keep this simple.

File: cmd/api/main.go [hunk 3]
```diff
@@ -1,37 +1,92 @@
...
+	signal.Notify(quit, os.Interrupt, os.Kill)
+	go func() {
+		<-quit
+		log.Println("server is shutting down...")
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+		if err := httpServer.Shutdown(ctx); err != nil {
+			log.Fatalf("could not shutdown gracefully: %v\n", err)
+		}
+		close(down)
```
Comment (christgf) [inline cmd/api/main.go]: Could work without the "server down" signal, but I suspect we'll want to wrap the http server and move it of here at some point, simplify main. The signal would definitely be more useful then.

File: cmd/api/main.go [hunk 4]
```diff
@@ -1,37 +1,92 @@
...
+			http.Error(w, "internal server error", http.StatusInternalServerError)
+		}
+		return
+	})
+}
+
 func main() {
+	// load configuration
+	listenAddr := ":8080"
+	if val, ok := os.LookupEnv("LISTEN_ADDR"); ok {
```
Comment (christgf) [inline cmd/api/main.go]: I believe this was "LISTENON"? I prefer "LISTEN_ADDR", but I can revert as it could break production :)

File: cmd/api/main.go (hunk 2 above)
Comment (yavosh) [reply cmd/api/main.go]: we need a task for logging, I use logrus in other places

File: cmd/api/main.go (hunk 1 above)
Comment (yavosh) [reply cmd/api/main.go]: ok

File: cmd/api/main.go (hunk 3 above)
Comment (yavosh) [reply cmd/api/main.go]: no strong feelings looks good

File: cmd/api/main.go (hunk 4 above)
Comment (yavosh) [reply cmd/api/main.go]: now you are getting picky
